sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java

```diff
@@ -685,74 +685,88 @@ protected Optional<IDataAdapter> findMatch(DoTypeName doTypeName, DaTypeName daT
+     * Be careful, this method does not check that the given rDtt is allowed by the lnType of this LN/LN0.
+     * It does not even check if rDtt exists in DataTypeTemplate section.
+     * That means that it will create the missing DOI/SDI/DAI, even if it is not consistent with DataTypeTemplate section.
+     * It is the caller responsibility to ensure the consistency between the given rDtt and the lnType of this LN/LN0 (which points to DataTypeTemplate section).
+     * See 9.3.5 "LN0 and other Logical Nodes" of IEC 61850-6.
+     * If given rDtt.isUpdatable() is false, the method does nothing (it will not check if DA/BDA is updatable in DataTypeTemplate section).
+     * This method will not remove any Val, it will only create new Val or replace existing Val (for example if rDtt.getDaName().getDaiValues() is empty, it does nothing)
+     *
+     * @param rDtt summarized Data Type Temple containing new DA val
+     * @throws ScdException when given rDtt is missing DoName or DaName
      */
     public void updateDAI(@NonNull ResumedDataTemplate rDtt) throws ScdException {
 
         if (!rDtt.isDoNameDefined() || !rDtt.isDaNameDefined()) {
             throw new ScdException("Cannot update undefined DAI");
         }
-        DoTypeName doTypeName = rDtt.getDoName();
-        DaTypeName daTypeName = rDtt.getDaName();
+        if (!rDtt.isUpdatable() || rDtt.getDaName().getDaiValues().isEmpty()) {
+            return;
+        }
+
+        AbstractDAIAdapter<?> daiAdapter = createDaiAdapterIfNotExists(rDtt.getDoName(), rDtt.getDaName(), rDtt.isUpdatable());
+        daiAdapter.update(rDtt.getDaName().getDaiValues());
+    }
 
+    /**
+     * Return the DAI (in DOI/SDI/DAI chain) matching the givens doTypeName and daTypeName.
+     * If it does not exist, create the missing DOI/SDI/DAI elements in this LN/LN0 if needed, based on the given doTypeName, daTypeName, and setting DAI.valImport to given isValImport.
+     * Be careful, this method does not check that the given rDtt is allowed by the lnType of this LN/LN0.
+     * It does not even check if rDtt exists in DataTypeTemplate section.
+     * That means that it will create the missing DOI/SDI/DAI, even if it is not consistent with DataTypeTemplate section.
+     * It is the caller responsibility to ensure the consistency between the given rDtt and the lnType of this LN/LN0 (which points to DataTypeTemplate section).
+     * See 9.3.5 "LN0 and other Logical Nodes" of IEC 61850-6.
+     * @param doTypeName DO to match with DOI and SDO (if any) to match SDI
+     * @param daTypeName DA to match DAI when no BDA, otherwise DA and BDA to match SDI and DAI.
+     * @param setValImportOnCreate when this method creates the DAI, it will set DAI.valImport attribute with this parameter
+     * @return adapter for existing DAI or created DAI.
+     */
+    private AbstractDAIAdapter<?> createDaiAdapterIfNotExists(DoTypeName doTypeName, DaTypeName daTypeName, boolean setValImportOnCreate) {
```

Comment:
Make it public to facilitate the testing of various scenarios.

Reply:
I'm not really agree, 
if this method is dedicated to be private then it should be private.  
Tests can be readable by other ways

Reply:
@massifben I think you plan to update this PR by this branch **241-updatedai-val-does-not-produce-subelements-sdi-as-expected_2** where no challenging class exit like DAITracker.

Reply:
I rebased `241-updatedai-val-does-not-produce-subelements-sdi-as-expected` on `241-updatedai-val-does-not-produce-subelements-sdi-as-expected_2` and forced push. So the PR is up to date.

Reply:
So we agree to keep the method private. I will test it through `updateDai`

> I'm not really agree, if this method is dedicated to be private then it should be private. Tests can be readable by other ways

So I keep it private (I need to adapt the the tests before closing this thread)